File: console/src/api/projects.ts
import { api } from "./client";
import { pageQuery, type ListMeta, type ListParams, type Page } from "./pagination";

export interface Project {
  id: string;
  name: string;
  description?: string;
  status: string;
  registration_policy?: string;
  oauth_allowed_redirect_urls?: string[];
  created_at: string;
  updated_at: string;
}

export interface ListProjectsResponse {
  projects: Project[];
  meta?: { total_count?: number; page_size?: number };
}

// 服务端项目列表（in-memory crud，默认 page_size=50，created_at DESC）；
// 对接服务端分页，pageSize 必传。
export async function listProjects(params: ListParams): Promise<Page<Project>> {
  const res = await api.get<ListProjectsResponse & ListMeta>("/server/projects", {
    params: pageQuery(params),
  });
  return { rows: res.data.projects ?? [], nextPageToken: res.data.meta?.next_page_token };
}

export async function getProject(id: string): Promise<Project> {
  const res = await api.get<Project>(`/server/projects/${id}`);
  return res.data;
}

export async function createProject(input: {
  id: string;
  name: string;
  description?: string;
}): Promise<Project> {
  const res = await api.post<Project>("/server/projects", input);
  return res.data;
}

export async function updateProject(
  id: string,
  input: { name?: string; description?: string; registration_policy?: string }
): Promise<Project> {
  const res = await api.patch<Project>(`/server/projects/${id}`, input);
  return res.data;
}

export async function deleteProject(id: string): Promise<void> {
  await api.delete(`/server/projects/${id}`);
}

// ---- OAuth 重定向白名单（settings auth.oauth_allowed_redirect_urls）----

export async function updateOAuthRedirectAllowlist(
  projectId: string,
  urls: string[]
): Promise<Project> {
  const res = await api.put<Project>(
    `/server/projects/${projectId}/oauth-redirect-allowlist`,
    { urls }
  );
  return res.data;
}

// ---- 邀请码（T-03，invite_only 注册策略）----

export interface InviteCode {
  id: string;
  project_id: string;
  code: string;
  max_uses: number;
  used_count: number;
  expire_at?: string;
  revoked: boolean;
  created_by: string;
  created_at: string;
}

export interface ListInviteCodesResponse {
  invite_codes: InviteCode[];
  meta?: ListMeta["meta"] & { total_count?: number };
}

// 邀请码列表（offset 型分页；服务端 clamp 默认 50 / max 100，created_at DESC）。
// 2026-09-22 修复：此前服务端忽略 page_size/page_token 固定返回前 100 条。
export async function listInviteCodes(
  projectId: string,
  params: ListParams
): Promise<Page<InviteCode>> {
  const res = await api.get<ListInviteCodesResponse>(
    `/server/projects/${projectId}/invite-codes`,
    { params: pageQuery(params) }
  );
  return { rows: res.data.invite_codes ?? [], nextPageToken: res.data.meta?.next_page_token };
}

export async function createInviteCode(
  projectId: string,
  input: { max_uses?: number; expire_at?: string }
): Promise<InviteCode> {
  const res = await api.post<InviteCode>(
    `/server/projects/${projectId}/invite-codes`,
    input
  );
  return res.data;
}

export async function deleteInviteCode(
  projectId: string,
  id: string
): Promise<void> {
  await api.delete(`/server/projects/${projectId}/invite-codes/${id}`);
}

File: internal/api/servergrpc/projects.go
package servergrpc

import (
	"context"

	serverv1 "github.com/torchwooddev/torchwood/genproto/server/v1"
	sharedv1 "github.com/torchwooddev/torchwood/genproto/shared/v1"
	appserver "github.com/torchwooddev/torchwood/internal/app/server"
	"github.com/torchwooddev/torchwood/internal/domain/projects"
	"github.com/torchwooddev/torchwood/internal/pkg/contexts"
	"github.com/torchwooddev/torchwood/pkg/crud"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/timestamppb"
)

type ProjectsService struct {
	serverv1.UnimplementedProjectsServiceServer
	projects *appserver.Projects
	invites  *appserver.InviteCodes
}

func NewProjectsService(projects *appserver.Projects, invites *appserver.InviteCodes) *ProjectsService {
	return &ProjectsService{projects: projects, invites: invites}
}

func (s *ProjectsService) CreateProject(ctx context.Context, req *serverv1.CreateProjectRequest) (*serverv1.Project, error) {
	p, err := s.projects.CreateProject(ctx, appserver.CreateProjectCommand{
		ID:          req.GetId(),
		Name:        req.GetName(),
		Description: req.GetDescription(),
	})
	if err != nil {
		return nil, err
	}
	return mapProject(p), nil
}

func (s *ProjectsService) ListProjects(ctx context.Context, req *sharedv1.ListRequest) (*serverv1.ListProjectsResponse, error) {
	list, info, err := s.projects.ListProjects(ctx, req.GetPageSize(), req.GetPageToken())
	if err != nil {
		return nil, err
	}
	var nextToken, prevToken string
	if info.HasNext {
		if nextToken, err = crud.EncodePageToken(info.NextOffset); err != nil {
			return nil, status.Error(codes.Internal, err.Error())
		}
	}
	if info.HasPrevious {
		if prevToken, err = crud.EncodePageToken(info.PreviousOffset); err != nil {
			return nil, status.Error(codes.Internal, err.Error())
		}
	}
	resp := &serverv1.ListProjectsResponse{
		Projects: make([]*serverv1.Project, len(list)),
		Meta: &sharedv1.ListResponseMeta{
			PageSize:      info.PageSize,
			NextPageToken: nextToken,
			PrevPageToken: prevToken,
			TotalCount:    int32(info.TotalCount),
		},
	}
	for i, p := range list {
		resp.Projects[i] = mapProject(&p)
	}
	return resp, nil
}

func (s *ProjectsService) GetProject(ctx context.Context, req *serverv1.GetProjectRequest) (*serverv1.Project, error) {
	p, err := s.projects.GetProject(ctx, req.GetId())
	if err != nil {
		return nil, err
	}
	if p == nil {
		// 对齐 GetUser 模式：use-case 返回 nil,nil 时显式转 NotFound，
		// 避免 gRPC OK + 空响应的错误分类。
		return nil, status.Error(codes.NotFound, "project not found")
	}
	return mapProject(p), nil
}

func (s *ProjectsService) UpdateProject(ctx context.Context, req *serverv1.UpdateProjectRequest) (*serverv1.Project, error) {
	ctx = contexts.WithAuditResource(ctx, req.GetId())
	cmd := appserver.UpdateProjectCommand{ProjectID: req.GetId()}
	if req.Name != nil {
		cmd.Name = req.Name
	}
	if req.Description != nil {
		cmd.Description = req.Description
	}
	if req.RegistrationPolicy != nil {
		cmd.RegistrationPolicy = req.RegistrationPolicy
	}
	p, err := s.projects.UpdateProject(ctx, cmd)
	if err != nil {
		return nil, err
	}
	return mapProject(p), nil
}

func (s *ProjectsService) DeleteProject(ctx context.Context, req *serverv1.GetProjectRequest) (*sharedv1.Empty, error) {
	ctx = contexts.WithAuditResource(ctx, req.GetId())
	if err := s.projects.DeleteProject(ctx, req.GetId()); err != nil {
		return nil, err
	}
	return &sharedv1.Empty{}, nil
}

// UpdateOAuthRedirectAllowlist 更新项目重定向白名单（整表替换；空列表 =
// 清空回落默认白名单）。
func (s *ProjectsService) UpdateOAuthRedirectAllowlist(ctx context.Context, req *serverv1.UpdateOAuthRedirectAllowlistRequest) (*serverv1.Project, error) {
	ctx = contexts.WithAuditResource(ctx, req.GetProjectId()+"/oauth-redirect-allowlist")
	p, err := s.projects.UpdateOAuthRedirectAllowlist(ctx, appserver.UpdateOAuthRedirectAllowlistCommand{
		ProjectID: req.GetProjectId(),
		URLs:      req.GetUrls(),
	})
	if err != nil {
		return nil, err
	}
	return mapProject(p), nil
}

func mapProject(p *projects.Project) *serverv1.Project {
	if p == nil {
		return nil
	}
	return &serverv1.Project{
		Id:                       p.ID,
		Name:                     p.Name,
		Description:              p.Description,
		Status:                   p.Status,
		RegistrationPolicy:       p.RegistrationPolicy,
		OauthAllowedRedirectUrls: projects.OAuthAllowedRedirectURLs(p.Settings),
		CreatedAt:                timestamppb.New(p.CreatedAt),
		UpdatedAt:                timestamppb.New(p.UpdatedAt),
	}
}

// ---- 邀请码（T-03）----

func (s *ProjectsService) CreateInviteCode(ctx context.Context, req *serverv1.CreateInviteCodeRequest) (*serverv1.InviteCode, error) {
	ctx = contexts.WithAuditResource(ctx, req.GetProjectId()+"/invite-codes")
	cmd := appserver.CreateInviteCodeCommand{ProjectID: req.GetProjectId()}
	if req.MaxUses != nil {
		cmd.MaxUses = req.MaxUses
	}
	if ts := req.GetExpireAt(); ts != nil {
		t := ts.AsTime()
		cmd.ExpireAt = &t
	}
	code, err := s.invites.Create(ctx, cmd)
	if err != nil {
		return nil, err
	}
	return mapInviteCode(code), nil
}

func (s *ProjectsService) ListInviteCodes(ctx context.Context, req *serverv1.ListInviteCodesRequest) (*serverv1.ListInviteCodesResponse, error) {
	codes, err := s.invites.List(ctx, req.GetProjectId())
	if err != nil {
		return nil, err
	}
	out := make([]*serverv1.InviteCode, len(codes))
	for i := range codes {
		out[i] = mapInviteCode(&codes[i])
	}
	return &serverv1.ListInviteCodesResponse{
		InviteCodes: out,
		Meta:        &sharedv1.ListResponseMeta{PageSize: int32(len(out))},
	}, nil
}

func (s *ProjectsService) DeleteInviteCode(ctx context.Context, req *serverv1.DeleteInviteCodeRequest) (*sharedv1.Empty, error) {
	ctx = contexts.WithAuditResource(ctx, req.GetProjectId()+"/invite-codes/"+req.GetId())
	if err := s.invites.Delete(ctx, req.GetProjectId(), req.GetId()); err != nil {
		return nil, err
	}
	return &sharedv1.Empty{}, nil
}

func mapInviteCode(c *projects.InviteCode) *serverv1.InviteCode {
	if c == nil {
		return nil
	}
	out := &serverv1.InviteCode{
		Id:        c.ID,
		ProjectId: c.ProjectID,
		Code:      c.Code,
		MaxUses:   int32(c.MaxUses),
		UsedCount: int32(c.UsedCount),
		Revoked:   c.Revoked(),
		CreatedBy: c.CreatedBy,
		CreatedAt: timestamppb.New(c.CreatedAt),
	}
	if c.ExpireAt != nil {
		out.ExpireAt = timestamppb.New(*c.ExpireAt)
	}
	return out
}
